expression/builtin_op.go

```diff
@@ -799,7 +799,11 @@ func (c *isNullFunctionClass) getFunction(ctx sessionctx.Context, args []Express
 		sig = &builtinRealIsNullSig{bf}
 		sig.setPbCode(tipb.ScalarFuncSig_RealIsNull)
 	case types.ETDatetime:
-		sig = &builtinTimeIsNullSig{bf}
+		isNotNull := false
+		if mysql.HasNotNullFlag(args[0].GetType().Flag) {
```

Comment:
Is it possible that the `NotNull` flag is generated from expressions, for example `date(a)`? Should we also specially handle it?

Reply:
How about add a more condition that only when `args[0]` is `Column` we can set `isNotNull` to ture?

Reply:
```
mysql> select c from t;
Field   1:  `c`
Catalog:    `def`
Database:   `test`
Table:      `t`
Org_table:  `t`
Type:       LONG
Collation:  binary (63)
Length:     11
Max_length: 1
Decimals:   0
Flags:      NOT_NULL NO_DEFAULT_VALUE NUM 


+---+
| c |
+---+
| 0 |
+---+
1 row in set (0.00 sec)

mysql> select date(c) from t;
Field   1:  `date(c)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATE
Collation:  binary (63)
Length:     10
Max_length: 10
Decimals:   0
Flags:      BINARY 


+------------+
| date(c)    |
+------------+
| 0000-00-00 |
+------------+
1 row in set (0.00 sec)
```

Looks like expression clears the not null flag?

Reply:
I think it is ok.

Reply:
@breeswish 
Try this case.
```
create table t (a datetime not null);
insert ignore into t values (0);
select * from t where a is null;
select * from t where date(a) is null;
```

In MySQL, the latter query returns nothing.

Reply:
Yes. I mean, it looks like `a` is NOT NULL does not mean that `date(a)` is NOT NULL, as indicated in https://github.com/pingcap/tidb/pull/10957#discussion_r298932689 . So this does not mean that this special rule applies only to columns.

Reply:
Actually looks like there is another TiDB bug..

```sql
create table t (a datetime not null);
insert ignore into t values (0);
select date(a) from t;
```

TiDB:

```
+---------+
| date(a) |
+---------+
| NULL    |
+---------+
1 row in set, 1 warning (0.01 sec)
```

MySQL:

```
+------------+
| date(a)    |
+------------+
| 0000-00-00 |
+------------+
1 row in set (0.00 sec)
```

Also the result of TiDB also indicates that `date(a)` can be NULL even if `a` is NOT NULL..

Reply:
There are lots of bugs(or differences) between TiDB and MySQL caused by different implementation details about type system and scalar functions...

Reply:
There may be another scalar function `f(a)` which can propagate the `NOT NULL` attribution from `a` to its outer world.
Check expression type and make it works only when it is a `Column` is safer.

Reply:
This column check may not work in TiKV..

Reply:
You will have no way to know the underlying expression is a column or not.

Reply:
Is this comment can be resolved?
I will fix the check problem in TiKV later.